### PR TITLE
Add debug_expr/2 and debug_expr_apply/3

### PR DESCRIPTION
**Pull request by josevalim**

### CODE REVIEW

File: nx/lib/nx/defn.ex [hunk 1]
```diff
@@ -443,6 +443,34 @@ defmodule Nx.Defn do
     end
   end
 
+  @doc """
+  Wraps an anonymous function to return its underlying defn expression.
+
+  This function must be invoked only for debugging purposes only.
```
Comment (jonatanklosko) [inline nx/lib/nx/defn.ex:449]:
```suggestion
  This function must be invoked for debugging purposes only.
```

We could also make it an admonition to bring more attention.

File: nx/lib/nx/defn.ex [hunk 2]
```diff
@@ -443,6 +443,34 @@ defmodule Nx.Defn do
...
+
+  """
+  def debug_expr(fun, opts \\ []) when is_function(fun) and is_list(opts) do
+    wrap(fun, &debug_expr_apply(fun, &1, opts))
+  end
+
+  @doc """
+  Invokes the anonymous function to return its underlying defn expression.
+
+  This function must be invoked only for debugging purposes only.
```
Comment (jonatanklosko) [inline nx/lib/nx/defn.ex:463]:
```suggestion
  This function must be invoked for debugging purposes only.
```

File: nx/lib/nx/defn/debug.ex [hunk 3]
```diff
@@ -1,4 +1,4 @@
-defmodule Nx.Defn.Identity do
+defmodule Nx.Defn.Debug do
   @behaviour Nx.Defn.Compiler
```
Comment (jonatanklosko) [inline nx/lib/nx/defn/debug.ex:2]:
```suggestion
defmodule Nx.Defn.Debug do
  @moduledoc false

  @behaviour Nx.Defn.Compiler
```

**Review by jonatanklosko — APPROVED** [review]

File: nx/lib/nx/defn.ex [hunk 4]
```diff
@@ -443,6 +443,34 @@ defmodule Nx.Defn do
     end
   end
 
+  @doc """
+  Wraps an anonymous function to return its underlying defn expression.
+
+  This function must be invoked for debugging purposes only.
```
Comment (josevalim) [inline nx/lib/nx/defn.ex:449]:
```suggestion
  > #### Warning {: .warning}
  >
  > This function must be invoked for debugging purposes only.
```

File: nx/lib/nx/defn.ex [hunk 5]
```diff
@@ -443,6 +443,34 @@ defmodule Nx.Defn do
...
+
+  """
+  def debug_expr(fun, opts \\ []) when is_function(fun) and is_list(opts) do
+    wrap(fun, &debug_expr_apply(fun, &1, opts))
+  end
+
+  @doc """
+  Invokes the anonymous function to return its underlying defn expression.
+
+  This function must be invoked for debugging purposes only.
```
Comment (josevalim) [inline nx/lib/nx/defn.ex:463]:
```suggestion
  > #### Warning {: .warning}
  >
  > This function must be invoked for debugging purposes only.
```

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Pull request merged**